In [1]:
import pandas as pd
import json

# Convertendo arquivos JSON para CSV

In [11]:
# Convertendo o arquivo JSON "clientes_crm" para o arquivo csv "customers_raw"

with open(r'JSON\clientes_crm.json', 'r', encoding='utf-8') as f:
    data = json.load(f)

dfcustomers = pd.DataFrame(data)

dfcustomers.to_csv('customers_raw.csv', index=False)

In [12]:
# Checando o dataframe gerado

dfcustomers.head(10)

,full_name,location,code,email
0,Femininos Oliveira Antunes,"Aratu (Candeias) , BA",1,femininos.oliveira.antunes@icloud.com
1,Fernanda Azevedo Soares Nunes Vieira,"PE , Recife",2,nunes.fernanda.soares.azevedo.vieira@outlook.com
2,Daniel Farias Ribeiro Teixeira,"Rio Grande,RS",3,farias.teixeira.daniel.ribeiro#gmail.com
3,Thiago Moreira,"AC , Rio Branco",4,thiago.moreira#gmail.com
4,Pedro Freitas,PA - Santarém Novo,5,pedro.freitas#icloud.com
5,Antônia Coelho Pinheiro Peixoto Cavalcanti,"Fortaleza do Tabocão , TO",6,coelho.pinheiro.peixoto.antônia.cavalcanti@aol...
6,Bianca Barros Rocha Torres Siqueira,PB/Cabedelo,7,torres.barros.rocha.bianca.siqueira#aol.com
7,Luiz Alves Pimentel,SE - Aracaju,8,pimentel.alves.luiz#outlook.com
8,Lucas Guedes Cunha Lopes,PB - João Pessoa,9,lucas.lopes.guedes.cunha#tutanota.com
9,Débora Paiva,Santarém / PA,10,paiva.débora#gmx.com


In [13]:
dfcustomers.dtypes

full_name    object
location     object
code          int64
email        object
dtype: object

In [14]:
dfcustomers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49 entries, 0 to 48
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   full_name  49 non-null     object
 1   location   49 non-null     object
 2   code       49 non-null     int64 
 3   email      49 non-null     object
dtypes: int64(1), object(3)
memory usage: 1.7+ KB


In [15]:
len(dfcustomers)

49

In [16]:
dfcustomers.isnull().sum()

full_name    0
location     0
code         0
email        0
dtype: int64

In [17]:
dfcustomers.duplicated().sum()

0

In [18]:
# Checando o csv gerado

dfcustomers_check = pd.read_csv('customers_raw.csv')

dfcustomers_check.head()
dfcustomers_check.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49 entries, 0 to 48
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   full_name  49 non-null     object
 1   location   49 non-null     object
 2   code       49 non-null     int64 
 3   email      49 non-null     object
dtypes: int64(1), object(3)
memory usage: 1.7+ KB


In [24]:
# Convertendo o arquivo JSON "custos_importacao" para o arquivo csv "costs_raw"

rows = []

with open(r'JSON\custos_importacao.json', 'r', encoding='utf-8') as f:
    data = json.load(f)

for product in data:
    for hist in product['historic_data']:
        rows.append({
            'product_id': product['product_id'],
            'product_name': product['product_name'],
            'category': product['category'],
            'start_date': hist['start_date'],
            'usd_price': hist['usd_price']
        })

dfcosts = pd.DataFrame(rows)

dfcosts.to_csv('costs_raw.csv', index=False)

In [25]:
# Checando o dataframe gerado

dfcosts.head(10)

,product_id,product_name,category,start_date,usd_price
0,1,Transponder AIS Maré Magnum,eletrônicos,10/08/2016,10583.63
1,1,Transponder AIS Maré Magnum,eletrônicos,15/06/2018,8778.36
2,1,Transponder AIS Maré Magnum,eletrônicos,25/09/2018,8023.87
3,1,Transponder AIS Maré Magnum,eletrônicos,19/03/2019,8772.78
4,1,Transponder AIS Maré Magnum,eletrônicos,17/01/2020,7918.18
5,1,Transponder AIS Maré Magnum,eletrônicos,17/06/2020,6310.01
6,1,Transponder AIS Maré Magnum,eletrônicos,02/07/2021,6586.70
7,1,Transponder AIS Maré Magnum,eletrônicos,16/05/2022,6538.20
8,1,Transponder AIS Maré Magnum,eletrônicos,28/02/2023,6360.91
9,1,Transponder AIS Maré Magnum,eletrônicos,17/10/2023,6574.80


In [27]:
dfcosts.dtypes

product_id        int64
product_name     object
category         object
start_date       object
usd_price       float64
dtype: object

In [28]:
dfcosts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1260 entries, 0 to 1259
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_id    1260 non-null   int64  
 1   product_name  1260 non-null   object 
 2   category      1260 non-null   object 
 3   start_date    1260 non-null   object 
 4   usd_price     1260 non-null   float64
dtypes: float64(1), int64(1), object(3)
memory usage: 49.3+ KB


In [29]:
len(dfcosts)

1260

In [30]:
dfcosts.isnull().sum()

product_id      0
product_name    0
category        0
start_date      0
usd_price       0
dtype: int64

In [31]:
dfcosts.duplicated().sum()

0

In [33]:
# Checando o csv gerado

dfcosts_check = pd.read_csv('costs_raw.csv')

dfcosts_check.head()
dfcosts_check.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1260 entries, 0 to 1259
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_id    1260 non-null   int64  
 1   product_name  1260 non-null   object 
 2   category      1260 non-null   object 
 3   start_date    1260 non-null   object 
 4   usd_price     1260 non-null   float64
dtypes: float64(1), int64(1), object(3)
memory usage: 49.3+ KB


# Checagem inicial de problemas

In [41]:
# Checando o CSV "produtos_raw"

df_prod = pd.read_csv('produtos_raw.csv')

df_prod.head()
df_prod.info()
df_prod.isnull().sum()
df_prod.duplicated().sum()
df_prod['actual_category'].value_counts().head(20)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 157 entries, 0 to 156
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   name             157 non-null    object
 1   price            157 non-null    object
 2   code             157 non-null    int64 
 3   actual_category  157 non-null    object
dtypes: int64(1), object(3)
memory usage: 5.0+ KB


actual_category
AncorageM                9
Ancoraguem               8
Propução                 8
Eletronicoz              7
eletrônicos              7
ELETRONICOS              6
PROPULSAO                6
E L E T R Ô N I C O S    6
P R O P U L S Ã O        6
propulsão                6
Encoragem                5
Propulção                5
Prop                     5
Propulssão               5
aNcOrAgEm                5
A N C O R A G E M        5
Eletrunicos              5
Ancorajm                 5
eLeTrÔnIcOs              5
propulsao                4
Name: count, dtype: int64

**Problema**: Na prática, apenas 3 categorias. Escritas de diversas maneiras, sem padronização.

In [42]:
df_prod[df_prod.duplicated()]

,name,price,code,actual_category


In [43]:
df_prod['name'].value_counts().head(10)

name
Motor Diesel Yanmar Velocity 37HP                4
Boia de Arqueamento Delta Nexus                  3
GPS Lowrance Evo Storm Drift                     2
Cabo de Nylon Delta Velocity Core Mako           2
Transponder AIS Maré Magnum                      1
Âncora Delta Swift                               1
Motor Diesel Parsun Velocity Abyss Orca 235HP    1
Motor Elétrico Honda Flux Ultra 273HP            1
Motor Elétrico Honda Vector Abyss 220HP          1
Cabo de Nylon Delta Vortex                       1
Name: count, dtype: int64

**Problema**: Nomes de produtos duplicados. Não aparecem como linha duplicada pois apresentam diferenciação na forma como a categoria está descrita na última coluna.

In [36]:
df_sales = pd.read_csv('vendas_2023_2024.csv')

df_sales.head()
df_sales.info()
df_sales.isnull().sum()
df_sales.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9895 entries, 0 to 9894
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id          9895 non-null   int64  
 1   id_client   9895 non-null   int64  
 2   id_product  9895 non-null   int64  
 3   qtd         9895 non-null   int64  
 4   total       9895 non-null   float64
 5   sale_date   9895 non-null   object 
dtypes: float64(1), int64(4), object(1)
memory usage: 464.0+ KB


0

**Problemas**:   
        1 - Apenas 9895 linhas não nulas, embora os IDs vão até 9999.   
        2 - Coluna de data em formato string.

In [37]:
# Checando datas

df_sales['sale_date'].head(20)

0     2023-09-10
1     15-09-2024
2     2024-08-13
3     2023-02-03
4     2024-02-12
5     2023-09-26
6     2024-02-28
7     07-11-2023
8     2024-08-25
9     2023-05-07
10    29-07-2024
11    28-04-2024
12    06-05-2024
13    2023-10-14
14    07-10-2024
15    2023-09-05
16    2024-11-28
17    2023-07-16
18    27-11-2024
19    04-12-2024
Name: sale_date, dtype: object

In [44]:
pd.to_datetime(df_sales['sale_date'], errors='coerce')

0      2023-09-10
1             NaT
2      2024-08-13
3      2023-02-03
4      2024-02-12
          ...    
9890          NaT
9891          NaT
9892          NaT
9893          NaT
9894          NaT
Name: sale_date, Length: 9895, dtype: datetime64[ns]

In [45]:
df_sales[pd.to_datetime(df_sales['sale_date'], errors='coerce').isna()]

,id,id_client,id_product,qtd,total,sale_date
1,1,3,136,9,16873.90,15-09-2024
7,9,37,26,7,114932.90,07-11-2023
10,12,27,48,13,191573.00,29-07-2024
11,13,43,19,2,31221.75,28-04-2024
12,14,13,125,15,48479.45,06-05-2024
...,...,...,...,...,...,...
9890,9995,30,139,6,8549.00,11-03-2023
9891,9996,9,111,7,28497.15,17-09-2023
9892,9997,38,123,2,5276.30,20-06-2023
9893,9998,33,97,6,771409.50,23-10-2024


**Problema**: Apenas 4982 das 9894 (~50%) datas estão em um formato reconhecível padrão.

In [50]:
# Checando datas com o formato invertido

pd.to_datetime(df_sales['sale_date'], errors='coerce', dayfirst=True)

0      2023-10-09
1             NaT
2             NaT
3      2023-03-02
4      2024-12-02
          ...    
9890          NaT
9891          NaT
9892          NaT
9893          NaT
9894          NaT
Name: sale_date, Length: 9895, dtype: datetime64[ns]

In [47]:
df_sales[pd.to_datetime(df_sales['sale_date'], errors='coerce', dayfirst=True).isna()]

,id,id_client,id_product,qtd,total,sale_date
1,1,3,136,9,16873.90,15-09-2024
2,2,25,139,7,9475.30,2024-08-13
5,6,36,52,3,39056.40,2023-09-26
6,8,27,25,3,34560.05,2024-02-28
7,9,37,26,7,114932.90,07-11-2023
...,...,...,...,...,...,...
9890,9995,30,139,6,8549.00,11-03-2023
9891,9996,9,111,7,28497.15,17-09-2023
9892,9997,38,123,2,5276.30,20-06-2023
9893,9998,33,97,6,771409.50,23-10-2024


=> Aumento de 4982 para 7945 datas válidas

In [48]:
df_sales['sale_date'].sample(20)

7215    2023-12-08
4020    31-05-2023
1469    26-01-2023
7921    28-11-2023
2116    09-06-2024
7880    2023-11-14
1353    31-10-2024
7138    29-10-2023
3631    26-11-2024
3458    2023-10-31
6333    11-10-2024
2931    19-07-2023
3949    2024-02-13
3785    31-03-2023
5198    25-07-2023
6074    2024-05-24
173     11-03-2023
4723    2023-05-09
8285    19-05-2023
9683    2023-10-06
Name: sale_date, dtype: object

In [49]:
df_sales['sale_date'].str.len().value_counts()

sale_date
10    9895
Name: count, dtype: int64

=> Datas estão ou no formato YYYY-MM-DD ou DD-MM-YYYY

In [51]:
# Checando o CSV "customers_raw"

df_cust = pd.read_csv('customers_raw.csv')

df_cust.head()
df_cust.info()
df_cust.isnull().sum()
df_cust.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49 entries, 0 to 48
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   full_name  49 non-null     object
 1   location   49 non-null     object
 2   code       49 non-null     int64 
 3   email      49 non-null     object
dtypes: int64(1), object(3)
memory usage: 1.7+ KB


0

In [52]:
df_cust['location'].value_counts()

location
BA - Porto Seguro                 2
Aratu (Candeias) , BA             1
TO/Fortaleza do Tabocão           1
PR , Antonina                     1
CE,Fortaleza                      1
Itacoatiara - AM                  1
São Luís,MA                       1
Rio Grande/RS                     1
Fortaleza / CE                    1
João Pessoa / PB                  1
Belém/PA                          1
MA , Itaqui (São Luís)            1
PE / Suape (Ipojuca)              1
Manaus/AM                         1
Corumbá,MS                        1
BA / Aratu (Candeias)             1
Vila do Conde (Barcarena) - PA    1
MA/Itaqui (São Luís)              1
Paranaguá / PR                    1
Santarém,PA                       1
Macapá/AP                         1
RJ / Niterói                      1
PE - Suape (Ipojuca)              1
Salvador,BA                       1
ES / São Mateus                   1
PE , Recife                       1
PA / Santarém                     1
Rio Grande,RS      

**Problema**: Locais escritos nos mais diversos formatos, sem padronização. 

In [53]:
df_cust['email'].sample(15)

6          torres.barros.rocha.bianca.siqueira#aol.com
13             vieira.silva.amaral.gabriela@icloud.com
0                femininos.oliveira.antunes@icloud.com
9                                 paiva.débora#gmx.com
23      peixoto.carla.araújo.silva.azevedo#outlook.com
11                      rafael.pereira.barros#zoho.com
42                    farias.jéssica.cunha@outlook.com
22                    rodrigues.mendonça.bruna#aol.com
7                      pimentel.alves.luiz#outlook.com
34                       martins.priscila#tutanota.com
38                    oliveira.renata.siqueira@aol.com
27                         rodrigues.bianca@icloud.com
14           lopes.alves.pacheco.rocha.carla#yahoo.com
36    nunes.borges.teixeira.pimentel.jéssica#yahoo.com
48              costa.tavares.antunes.diego#icloud.com
Name: email, dtype: object

**Problema**: Alguns emails possuem um "#" ao invés de "@"

In [54]:
# Checando o CSV "costs_raw"

df_costs = pd.read_csv('costs_raw.csv')

df_costs.head()
df_costs.info()
df_costs.isnull().sum()
df_costs.duplicated().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1260 entries, 0 to 1259
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_id    1260 non-null   int64  
 1   product_name  1260 non-null   object 
 2   category      1260 non-null   object 
 3   start_date    1260 non-null   object 
 4   usd_price     1260 non-null   float64
dtypes: float64(1), int64(1), object(3)
memory usage: 49.3+ KB


0

In [55]:
# Checando datas

df_costs['start_date'].head(20)

0     10/08/2016
1     15/06/2018
2     25/09/2018
3     19/03/2019
4     17/01/2020
5     17/06/2020
6     02/07/2021
7     16/05/2022
8     28/02/2023
9     17/10/2023
10    16/02/2024
11    22/02/2024
12    15/03/2024
13    02/08/2024
14    08/04/2025
15    23/11/2017
16    06/07/2022
17    24/05/2023
18    05/11/2025
19    12/04/2016
Name: start_date, dtype: object

In [66]:
# Convertendo para o formato data

pd.to_datetime(df_costs['start_date'], errors='coerce')

0      2016-10-08
1             NaT
2             NaT
3             NaT
4             NaT
          ...    
1255   2023-01-06
1256          NaT
1257          NaT
1258   2024-04-10
1259          NaT
Name: start_date, Length: 1260, dtype: datetime64[ns]

In [57]:
df_costs[pd.to_datetime(df_costs['start_date'], errors='coerce').isna()]

,product_id,product_name,category,start_date,usd_price
1,1,Transponder AIS Maré Magnum,eletrônicos,15/06/2018,8778.36
2,1,Transponder AIS Maré Magnum,eletrônicos,25/09/2018,8023.87
3,1,Transponder AIS Maré Magnum,eletrônicos,19/03/2019,8772.78
4,1,Transponder AIS Maré Magnum,eletrônicos,17/01/2020,7918.18
5,1,Transponder AIS Maré Magnum,eletrônicos,17/06/2020,6310.01
...,...,...,...,...,...
1252,150,Cabo de Nylon Danforth Magnum Vox,ancoragem,19/02/2019,442.38
1253,150,Cabo de Nylon Danforth Magnum Vox,ancoragem,17/09/2020,312.94
1256,150,Cabo de Nylon Danforth Magnum Vox,ancoragem,31/01/2024,332.26
1257,150,Cabo de Nylon Danforth Magnum Vox,ancoragem,29/08/2024,292.03


=> 801 das 1259 datas não convertidas

In [67]:
# Checando para formato dd/mm/yyyy

pd.to_datetime(df_costs['start_date'], dayfirst=True, errors='coerce')

0      2016-08-10
1      2018-06-15
2      2018-09-25
3      2019-03-19
4      2020-01-17
          ...    
1255   2023-06-01
1256   2024-01-31
1257   2024-08-29
1258   2024-10-04
1259   2024-10-25
Name: start_date, Length: 1260, dtype: datetime64[ns]

In [64]:
df_costs[pd.to_datetime(df_costs['start_date'], dayfirst=True, errors='coerce').isna()]

,product_id,product_name,category,start_date,usd_price


=> Datas todas no formato brasileiro (dd/mm/yyyy)